### PR TITLE
Fix ModelHubMixin - kwargs should be passed correctly when reloading

### PR DESCRIPTION
**Pull request by Wauplin**
Fix https://github.com/huggingface/huggingface_hub/issues/2096 cc @not-lain :) 

In this PR:
- make all attributes related to ModelHubMixin internals private: `_hub_mixin_config`, `_hub_mixin_info`, `_hub_mixin_init_parameters`, etc. => fixes interference with the main class (cc @mfuntowicz). This is a breaking change since `self.config` doesn't exist anymore but not very annoying (was introduced in 0.20.x)
- if `config` is expected by `_from_pretrained` => pass it => fix for @mfuntowicz 
- all values in config are passed as kwargs if the class `__init__` expects kwargs => fix for @not-lain 

Not in this PR:
- we don't have a 100% coverage of possible use cases, especially if kwargs are passed to `__init__` but should not be passed to `_from_pretrained` => not our problem / we don't want to support that 

--- 

**EDIT:** I've completely changed this PR.
**EDIT 2:** I've again changed this PR.

Before the PR, if `**kwargs` was expected by the `__init__` method, then we would pass `config=config` by default to it, expecting the class not to fail if an unexpected kwarg is passed. This was to fix a breaking change introduced in 0.21.x reported by @mfuntowicz (see https://github.com/huggingface/huggingface_hub/pull/2058). However, this is to solve a corner case that is not much used. Instead of that, we should provide all kwarg values as kwargs. If `_from_pretrained` **really** want to see config injected, then it should expect `config` as arg. This is a very corner case and except the optimum-neuron example, I don't think we'll ever see it.

There is still a use case we don't handle and that is if the subclass expects some args including a `config` one and some kwargs. Then the config values will also be forwarded as kwargs which _might_ break initialization in case kwargs are again forwarded to another method. We don't support such a case and **that's on purpose**. It's best not to support weird cases and rather encourage authors not to configure their class with a config + some undocumented kwargs.

**This doesn't change the fact that the preferred way to deal with config is to explicitly expect `config` or any jsonable values in the `__init__` function.** This is the best way to guarantee the config values will be saved and reloaded automatically, without additional boilerplate code.  

---

In the current form, example from #2096 is fixed.

--- 

@mfuntowicz This PR is a breaking change for your use case. What you'll have to do is simply to expect `config` as an optional argument in `_from_pretrained` and that's it.

### CODE REVIEW

**Comment by Wauplin** [comment]
> pass the config_inject_mode to the config.json when we are calling the push_to_hub

I feel this is more painful to use. It means actual users of the library will have to set this flag when pushing or loading from the Hub. With the current PR, only library authors and users integrating a library with the Hub will have to understand this flag - which I think can be much easier to convince.

Also, do you have an example of library that requires a kwarg arguments but did not define it in their __init__ method? I feel that should be the corner case, not the main usage. Maybe it would help me grasp what's the problem with the current solution.

**Comment by not-lain** [comment]
i didn't use the inspect library much before, and i am not sure of the exact ins and outs around it, i'm sure it's even harder when we are trying to identify what the descendant class has from ancestor class, so if possible could i get from here till monday to check if there is another way around this without using letting people manually setup the `config_inject_mode` .

> Also, do you have an example of library that requires a kwarg arguments but did not define it in their init method? I feel that should be the corner case, not the main usage. Maybe it would help me grasp what's the problem with the current solution.

no on the opposite actually, i really love how this pr takes all the possible cases into consideration, just wanted to try to find out if it is possible to identify which `config_inject_mode`  we are falling into without letting people try to manually setting it up.

**Comment by NielsRogge** [comment]
I agree with @not-lain, requiring users to explicitly pass `config_inject_mode` is not desirable, and would ideally be abstracted way. Isn't it possible to automatically save the `kwargs` in the JSON as individual fields, just like this happens in the Transformers library?

e.g. the following works without issues:

```python
from transformers import BertConfig 

config = BertConfig(vocab_size=2, random_parameter=2)

config.save_pretrained(".")
```

This correctly serializes random_parameter in the config.json. BertConfig accepts **kwargs.

**Comment by Wauplin** [comment]
My point is: "how can we guess if the underlying model expects `config={"vocab_size": 2,  'random_param': 1...}` or `vocab_size=2, random_param=1, ...` to be passed when nothing is specified" (e.g. expects **kwargs). Some models expect the first one (e.g. examples in `optimum` from @mfuntowicz) while some other models expects the second one (e.g.your BertConfig example). 

So of course we could change the behavior in optimum but I guess some other models in the wild expects the same. It's not a problem of how to serialize the data in `config.json` since in both cases the config is the same. The problem is how to deserialize it correctly, in the context of the library. This `config_inject_mode` suggestion is a way to tackle this topic by allowing library authors (not end users) to specify how to inherit from ModelHubMixin (which is meant to be as generic as possible)

**Comment by Wauplin** [comment]
Just to clarify, we are only talking about cases where args are NOT explicit. If config is an arg then it will be passed correctly -already the case and will still be the case with this PR-. The ambiguous case is when **kwargs are expected.

**Comment by not-lain** [comment]
@Wauplin 
there is a corner case to this pr which is this 
```python
from torch import nn
from huggingface_hub import PyTorchModelHubMixin
class MyModel( nn.Module , PyTorchModelHubMixin ) : 
     def __init__(self, config , a , b , **kwargs ) : 
      super().__init__()
      assert a == "a" 
      assert b == "b" 
      assert kwargs["c"] == "c" 
      assert list(kwargs.keys()) == ["c"] # example Dust3r does not take everything as possible kwargs or it will generate an error
      self.layer = nn.Linear(config["a"], config["b"] ) 

config = {"a" : 1,
          "b" : 2 }
model = MyModel(config , a = "a" , b = "b" , c = "c" ) 
```
I mean the way we are giving config too much importance when we should've taken the easy way out and consider config as a normal parameter and give it's own key should be the case.

i made this [notebook](https://colab.research.google.com/drive/1tF2me7A7g-n-QnfFl5ulldJf3z6oVQGz?usp=sharing) that might help you understand what kind of serialization we should aim for, I also posted some additional info in slack

**Review by not-lain — CHANGES_REQUESTED** [review]
https://github.com/huggingface/huggingface_hub/pull/2099/files#diff-bf673df3e700e1f9ed9cffa6ce51441acd0474bb67f7ac7728aa336d43216285R283

could we change the name of `self.config` param since it can be overriden in __init__ (it's just a possibility and to avoid any broken cases)

**Comment by Wauplin** [comment]
I've made a last change in https://github.com/huggingface/huggingface_hub/pull/2099/commits/888781b074ecec757a1a204619484371780e239e and updated the PR accordingly. I completely removed `inject_config_mode` variable as it was too confusing. Now the default behavior is to pass values as kwargs when **kwargs is expected (e.g. #2096 is fixed).

And for @mfuntowicz the only change is to update `_from_pretrained` in optimum neuron which is a minimal change as well.

**Comment by not-lain** [comment]
this is waaaay better, nice work @Wauplin ✨

File: src/huggingface_hub/keras_mixin.py [hunk 1]
```diff
@@ -57,6 +57,7 @@ def _flatten_dict(dictionary, parent_key=""):
 
 def _create_hyperparameter_table(model):
     """Parse hyperparameter dictionary into a markdown table."""
+    table = None
```
Comment (LysandreJik) [inline src/huggingface_hub/keras_mixin.py:60]: cleaner :)

File: src/huggingface_hub/keras_mixin.py [hunk 2]
```diff
@@ -444,6 +443,7 @@ def _from_pretrained(
         resume_download,
         local_files_only,
         token,
+        config: Optional[Dict[str, Any]] = None,
```
Comment (LysandreJik) [inline src/huggingface_hub/keras_mixin.py:446]: If I'm not mistaken, PyTorch's mixin doesn't take a config as attribute there:

https://github.com/huggingface/huggingface_hub/blob/05585741416faf1ca5364da63c450ae56adbfb9b/src/huggingface_hub/hub_mixin.py#L598-L612

Is this a difference that we want to keep or do we want to have the two mixins behave the same?

**Review by LysandreJik — APPROVED** [review]
Ok, LGTM! A bit annoying for the `config` breaking change but I understand it wasn't really used until now.

Awesome work!

File: src/huggingface_hub/keras_mixin.py (hunk 2 above)
Comment (Wauplin) [reply src/huggingface_hub/keras_mixin.py:446]: Yes the difference is that `config` is injected to `_from_pretrained` only if the method expects it. In Pytorch mixin we don't necessarily need it (depends on `__init__`) while the Keras mixin expects it (to do `model.config = config` => not very useful but avoid a breaking change).

**Comment by Wauplin** [comment]
Thanks for the review! Let's get this merged now :tada:

**Pull request merged**